### DISCUSSION TITLE
How do we validate input data?

### DISCUSSION THREAD

**Discussion by chris48s** [badges/shields#5619] [Ideas]
See longer conversation in https://github.com/badges/shields/issues/5597#issuecomment-699542921

> What exactly is the purpose of the schema validation?

> How loose/strict they should be?

Some suggested principles for input validation:

* Our definition of "valid" should be no stricter than the upstream service's definition of "valid". Perhaps to phrase this in RFC-like language: "our validation MAY be less strict than the upstream service but MUST NOT be more strict than the upstream service". Exactly where in that grey are we fall is going to be determined by other badge/service-specific criteria/requirements.

* If theory (docs) and practice (real-world API responses) conflict, real-world outputs take precedence over documented behaviour. e.g: if the docs say `version` is a semver but we learn that there are real-world packages where the version number is `0.3b` or `1.2.1.27` then we should accept those values in preference to enforcing the documented API behaviour.

* Shields is descriptive ("the registry says your package version is 0.7, therefore your package version is 0.7") rather than prescriptive ("you're doing versioning wrong - lol"). We reflect the established norms of the communities we serve. If it's good enough for the upstream, it's good enough for us.

Some other thoughts on principles we can apply. Maybe these ones are more about how we identify when validation that already exists is incorrect rather than how we write schemas from scratch:

* If we know of a real-world example of a package/repo/etc that causes us to render an error value on a badge (e.g: ![](https://img.shields.io/badge/version-null-blue) ![](https://img.shields.io/badge/coverage-NaN%25-red) ![](https://img.shields.io/badge/build-undefined-red) etc), our input validation is broken and we should fix it.
* If we know of a real-world example of a package/repo/etc that causes us to thrown an unhandled runtime exception, our input validation is broken and we should fix it.
* We should not fail to render a badge because of a validation failure on a field that isn't necessary to render the badge. (for example, a version badge should not fail to render because the license field is missing). It is not inherently wrong to define a single schema which is applied to multiple badges. For example, its fine in principle to define a schema that says 

    ```js
    const schema = Joi.object({
      version: Joi.string().required(),
      license: Joi.string().required(),
    }).required()
    ```

    for a shared helper and have both the license and version badges validate the response against that schema. However, if we become aware of a real-world example of a package/repo/etc that has a `version` key but not a `license` key then we should split the schema (or make `version` optional and handle the error in code). Maybe we know that at schema creation time (e.g: because its documented), or maybe we don't find out until some later point in time (because we have a real-world example of it happening via a support request).

Would welcome input from other maintainers...

**Comment by calebcartwright** [comment]
Cross posting the gist of my last comment from the linked issue for reference


Still digesting and thinking on this one, but wanted to note that I look forward to the discussion and hopefully being able to decide on a direction. There's been different and yes some times conflicting prescriptions on the "how strict/loose" topic, and it was one of the more confusing, and occasionally frustrating, things for me when I first started contributing to Shields.

Ultimately I don't care all that much about what the decided outcome is, but really want to avoid contributors receiving different messages depending on which maintainer is reviewing (and what may be on that maintainer's mind that day), or which docs/existing badges are used as a reference.

**Comment by calebcartwright** [comment]
@chris48s - my takeaway theme from the guiding points you've articulated would be: 

> define all schemas as loose as they can possibly be without causing potential runtime type errors, and retroactively tighten them if/when needed.

I think in general I can get behind that strategy, although I do still think there's merit to having stricter schemas that surface incorrect assumptions by failing more loudly/obviously. I don't have a good example offhand, but my sense is that we've had our fair share of bugs that were more easily debugged due to the `invalid response data` error reported by the user.

The most important outcome of this discussion from my perspective is that we document the agreed upon strategy, and use/enforce that going forward (as well as adjusting existing non-compliant schemas over time). Perhaps I'm wrong, but I'd posit there's a mix of both ends of the schema strictness scale today and that there are plenty that are overly strict in the context of this proposal, and that they did not get there iteratively but were instead born very tight/strict.

**Comment by PyvesB** [comment]
Will have a think as well, but reading through your principles makes sense to me so far @chris48s !

**Comment by chris48s** [comment]
> my takeaway theme from the guiding points you've articulated would be:

> > define all schemas as loose as they can possibly be without causing potential runtime type errors, and retroactively tighten them if/when needed.

I wouldn't consider that a correct summary, or certainly not a complete one.

Rendering a badge that says ![](https://img.shields.io/badge/coverage---1%25-red) doesn't throw a runtime exception. We should still not render it. Part of what we're trying to is prevent unexpected or undesirable outputs/behaviours.

Also, if you think about a schema like

```js
license: Joi.string().required()
```

we could make that

```js
license: Joi.alternatives().try(Joi.string().required(), Joi.number().required())
```
That is more lenient (the second constraint will accept more inputs than the first and its not going to cause us to throw any more errors than the first one), but it would also make no sense and it does not fit in with out understanding of what a license is. The schema should be informed by (and explain) our understanding of the input data, not just be "as loose as they can possibly be".

Perhaps another good purpose to think of a schema as having is as well as preventing us from accepting invalid data, its also partly about explaining/documenting to a human reading this code what we expect the input data to look like.

The way I would characterise it more is that any particular service/API is going to have a "window of sensible-ness" within which our schema should sit and I've attempted to write the (non-exhaustive) start of some principles that help us to determine
a) where each end of that window is and
b) under what circumstance we need to re-evaluate the window
but as long as our schema is somewhere inside that window, other characteristics of the schema (like simplicity, or readability, or providing better documentation of our expectations about the input data) might also inform exactly exactly where we choose to sit within that window in a particular case.

> although I do still think there's merit to having stricter schemas that surface incorrect assumptions by failing more loudly/obviously

Another principle I've not made explicit but I'm going to suggest we encode is that the schema/validation we choose is informed by the assumptions we're making about the data. This can include basic type checks. For example: If we need to multiply it by something, we check its a number. If we're going to call `.split()` on it, we make sure its a string. If we're going to address `foo[0]`, `foo` must be an array. But this can include more nuanced stuff e.g: If we're going to sort a version on the assumption it is a semver, check its a semver. I agree we do want to know that an upstream API is returning non-semver compliant versions if we're passing them to `semver.compare()` (whether that will throw an exception or just produce an unexpected result) and I'm definitely not suggesting we should loosen a schema that asserts a characteristic we are actively relying on. That's a case where the format of the version _does_ matter. Equally though, if we're just going to bung `foo.latest_version` on a badge, we can dispense with a check of that nature. Check the stuff that matters, don't check the stuff that doesn't matter.

Does that cover that point?


> The most important outcome of this discussion from my perspective is that we document the agreed upon strategy, and use/enforce that going forward

The most important thing for me is that this doesn't just make a rod for our own backs, or turn into somehow an reason to not fix a bug when its clearly a bug. If this just turns into telling a user "![](https://img.shields.io/badge/version-null-blue) is correct because we wrote some stuff in a markdown document", this process is a waste of our time and everyone else's. Any decision or rule we write down should be amendable in light of better evidence. Documentation (like code) is not a sacred cow. It is just some stuff we made up, and our future selves can change it. They will be smarter than us.

**Reply by calebcartwright** [reply]
> The most important thing for me is that this doesn't just make a rod for our own backs, or turn into somehow an reason to not fix a bug when its clearly a bug. If this just turns into telling a user " is correct because we wrote some stuff in a markdown document", this process is a waste of our time and everyone else's. Any decision or rule we write down should be amendable in light of better evidence. Documentation (like code) is not a sacred cow. It is just some stuff we made up, and our future selves can change it. They will be smarter than us.

I'm sensing quite a bit of anger/frustration in this, and hopefully that's not due to anything I've said or the way I've said it (my apologies if so). All I'm asking for is that we have some guidance we can point contributors to around what to do when it comes to strictness of schemas. I don't think that's unreasonable, nor do I think it's a waste of time. 

I mentioned previously that I felt there was a lack of clarity on this topic when I was first contributing to Shields, and that's still a sense I have today. I don't think that experience/perception should be dismissed out of hand, even if others feel differently.

I'm not trying to define strategy nor documentation for the sake of doing it, nor do I strive to produce bad badges/`invalid response data` badges. I just wanted to have a discussion and agree to something instead of continuing the ad-hoc and often conflicting approaches we've taken to this recurring question.

**Reply by calebcartwright** [reply]
> I wouldn't consider that a correct summary, or certainly not a complete one.

Yeah that's fair. I got overly strict in my theme-defining schema :wink: 

> Does that cover that point?

Yes, I was just trying to distill it down to a one liner but lost some important nuance. I guess it' could be more of a.. no stricter than necessary to render a good badge, but completely agree that what meets that qualification of good/sensible-ness will absolutely be context/API/service specific.

**Comment by chris48s** [comment]
> I'm sensing quite a bit of anger/frustration in this

I thought the tone of this draft was fairly light-hearted tbh. Maybe it needed more emoji

> All I'm asking for is that we have some guidance we can point contributors to around what to do when it comes to strictness of schemas

Guidance is good. Writing down what we do and why we do it is helpful. I fully agree with this - hence having bothered to spend time on this exercise so far. I suspect I will spend a fair chunk more time on it before we're done with this.

Being too rigid about enforcing that guidance, not so much. I guarantee at least one of the things I've written here sounds reasonable now but will turn out to be wrong in the face of a sensible example. Feel free to be the person that provides it.

FWIW, I think this process should be mostly about describing the good practices we've learned from writing hundreds of these things and learning from our mistakes rather than defining a swathe of changes we're going to apply to make everything "compliant". Obviously I've not read every single one of the hundreds of schemas we've already defined and I'm sure there's some weird stuff in there, but I reckon the vast majority are already somewhere inside the "window of sensible-ness".

> I mentioned previously that I felt there was a lack of clarity on this topic when I was first contributing to Shields, and that's still a sense I have today

Anything in this thread helping with that, or not really?

> often conflicting approaches

Got any particular examples of that? If you can't bring any to mind, no worries, but if we've got some examples of the kind of situation we're trying to clarify or inconsistency we're trying to prevent then that is going to be a big help in writing the docs that are supposed to clarify it.

**Reply by calebcartwright** [reply]
> I thought the tone of this draft was fairly light-hearted tbh. Maybe it needed more emoji

No worries! I probably just read that section a bit too literally, just a case of misinterpretation of written text

> Anything in this thread helping with that, or not really?

Yup, getting there

> Got any particular examples of that? If you can't bring any to mind, no worries, but if we've got some examples of the kind of situation we're trying to clarify or inconsistency we're trying to prevent then that is going to be a big help in writing the docs that are supposed to clarify it.

We'd previously had discussions about the inconsistency, such as [here](https://github.com/badges/shields/pull/2660#discussion_r245530168) (ironically with unassigned actions to document things :laughing:). That thread actually has good examples around the impacts of strictness/looseness on the build status badges. The tradeoffs discussed there around more vs. less strict was a successfully rendered badge with status but an off color vs. failing loudly with an `invalid response data` and handling accordingly.

We have version badges that are very loose, like [amo](https://github.com/badges/shields/blob/ff9273a958e8570dcb64f593f51b7743a1ca551f/services/amo/amo-base.js#L12) and others that are _very_ strict like [nexus](https://github.com/badges/shields/blob/ff9273a958e8570dcb64f593f51b7743a1ca551f/services/nexus/nexus.service.js#L17) (which has definitely been a pain point as well). I can't recall a specific case offhand, but know there's been instances in PR reviews of reviewer asking for a tighter (or looser) schema.

**Reply by chris48s** [reply]
> We have version badges that are very loose, like amo and others that are very strict like nexus

I would say nexus is probably a good example of one where we are checking for something we don't really need to check at input time. It looks like nothing we are doing with `latestRelease` or `latestSnapshot` really requires them to be in a specific format.


> That thread actually has good examples around the impacts of strictness/looseness on the build status badges

Good example! (and also horrible example :scream: )
Build statuses is a category where there probably is inconsistency and where somewhat differing opinions exist within the team. Its also a category where we've got a lot of badges so worth talking about specifically. I think we do have consensus that not knowing what colour to render the badge is a condition we want to prevent and we should validate to guard against that. There's also the weirdness that we sometimes double-up `isBuildStatus` as an input validator as well as a test validator (history on how we started off down the path to https://github.com/badges/shields/blob/master/services/build-status.js if you haven't read it before: https://github.com/badges/shields/pull/1767 ).
This one probably needs us to review what we've already got, work out what's good and what's bad and maybe needs more viewpoints to form a consensus but I guess the first bit of homework for that one is to have a look at https://github.com/badges/shields/search?q=static+category+%3D+%27build%27 and see what we've got/where the outliers are.

**Reply by chris48s** [reply]
OK, so I've had a look at all our build status badges and put them into some categories. Just to put a scope on this, I'm only looking at "build status" badges i.e: ![](https://img.shields.io/badge/build-passing-brightgreen). There are some other badges in the "build" category like

![](http://img.shields.io/badge/docker%20build-automated-066da5) / ![](http://img.shields.io/badge/docker%20build-manual-066da5)
![](http://img.shields.io/badge/tests-20%20passed,%201%20failed,%202%20skipped-red)

Those are a different thing.

This exercise doesn't really _answer_ any questions, but it does hopefully break down "how do we do validation for build status badges" into a number of more specific questions.

The most common type of badge we've got are ones which use `isBuildStatus` to validate the inputs and `renderBuildStatusBadge()` to render the output. Within that there are 2 sub-categories.

- Ones that use `isBuildStatus` with an extra invalid status e.g:
    ```js
    const schema = Joi.object({
      status: Joi.alternatives().try(isBuildStatus, Joi.equal('unknown')),
    }).required()
    ```
- Ones that just use `isBuildStatus` on its own e.g:
    ```js
    const schema = Joi.array().items({
      result: isBuildStatus.required(),
    })
    ```

The list of services where we use `renderBuildStatusBadge()` and `isBuildStatus` with an extra invalid status are:

- Azure Devops Build
- Azure Devops Release
- Buildkite
- Cirrus CI
- CodeShip
- Continuous PHP
- Drone
- Github workflow status
- Gitlab pipeline status
- ReadTheDocs
- Scrutinizer
- Travis CI

..and then the ones where we just use `isBuildStatus` are:

- Appveyor
- Circle CI
- Visual Studio App Centre
- Wercker

Those two categories seem to be the majority/dominant approach. I guess the main question there is: are we happy with the way invalid status values are dealt with? (reason for why things are how they are: https://github.com/badges/shields/pull/1767#issuecomment-404379521 )

The other categories we have are:

Build status badges that use `renderBuildStatusBadge()` but _don't_ use `isBuildStatus`:

- Bitbucket Pipelines (custom schema)
- Jenkins (custom schema)
- Netlify (no schema at all?)
- Shippable (custom schema)

and build status badges that don't use either `renderBuildStatusBadge()` or `isBuildStatus`

- Bitrise (custom schema)
- Docker build (custom schema)
- Docker Cloud build (custom schema)
- Github Pull Request Check State (custom schema)
- Teamcity (custom schema)

I guess the questions we want to answer with those two categories are:

- Are these badges doing things differently for good reasons, or are they inconsistent for no reason?
- If there are cases where they are different for good reasons, what are those reasons and how do we document what the reasons are? I think its fine if some badges are different as long as we can explain why and identify those patterns elsewhere in future.
- If they are different for no good reason, can we bring them into line without making a behaviour change? If it does require a behaviour change, is that desirable? do we want to make it?

I haven't really attempted to answer those questions, but identifying them seems like a start...

**Reply by calebcartwright** [reply]
Thanks for all the detailed analysis!

> I guess the main question there is: are we happy with the way invalid status values are dealt with?

Just to clarify, is that in reference to the usage of `isBuildStatus` for validation with usage of `Joi.alternatives` to handle cases where there are additional values we may see from a service that don't warrant inclusion in `isBuildStatus`? If so then my answer is yes, I'm content with that.

I feel like a good and sensible default approach for build services (still excluding the test status badges that happen to be in the same category). Over time we may discover some additional values that make sense to use globally across all build service badges and accordingly add to `isBuildStatus`, but there will also likely be times when Service Foo has an additional status that would be service specific and potentially conflict with some other Service Bar so we don't want to include in `isBuildStatus`


> I guess the questions we want to answer with those two categories are:
> 
>     * Are these badges doing things differently for good reasons, or are they inconsistent for no reason?
> 
>     * If there are cases where they are different for good reasons, what are those reasons and how do we document what the reasons are? I think its fine if some badges are different as long as we can explain why and identify those patterns elsewhere in future.
> 
>     * If they are different for no good reason, can we bring them into line without making a behaviour change? If it does require a behaviour change, is that desirable? do we want to make it?

I think we'll probably have to answer these on a case by case basis. For example, the only difference with the Bitbucket Pipelines response is that the values are all upper case and IIRC our schema validation is case-sensitive, but we could probably change this to `isBuildStatus.insensitive()` or `isBuildStatus.uppercase()` and then call `.toLowerCase()` in the `transform` function

https://github.com/badges/shields/blob/0737e317daed816c98cf82ab1c0bb7f8ee8df78f/services/bitbucket/bitbucket-pipelines.service.js#L14-L19

Jenkins on the other hand returns response values that are actually colors :rainbow: which we have to map ourselves into a status value
https://github.com/badges/shields/blob/0737e317daed816c98cf82ab1c0bb7f8ee8df78f/services/jenkins/jenkins-build.service.js#L12-L34

**Reply by chris48s** [reply]
Yeah I think each one is going to be a bit different. I think I'd agree that bitbucket could probably be massaged into shape, whereas jenkins is probably a special unique snowflake :snowflake: (I think that's what you're getting at anyway) but I don't know how to turn that into a rule

**Reply by calebcartwright** [reply]
> but I don't know how to turn that into a rule

Let me know if I've misunderstood or am missing something, but the core guideline would still be the same right? It's more of specific guidance for build status badges to that end. 

So the whole, schemas should be no stricter than necessary to render a correct badge (message, color, etc.) guide still applies. For build status badges the additional guidance is to default to the `isBuildStatus` validator, and generally don't go outside that unless you really need to (e.g. Jenkins).

Personally I feel like we've just about got enough here to document our core guidelines on schemas. The main outstanding items in mind are:

1. Specific guidance for certain scenarios (this seems to be heading towards aligning to certain badge categories)
2. Stricter schema validation vs. looser validation with logic checks in the service class to provide more contextual error message
3. Whether/when to update existing badge schemas that diverge from the guidelines

**Reply by chris48s** [reply]
> default to the `isBuildStatus` validator, and generally don't go outside that unless you really need to

Yeah I'm happy to not be too specific, Maybe we're just starting from different points and discovering that `isBuildStatus` is the most prevalent validator so we can say "default to `isBuildStatus` if you can" is enough of a clarification at this point. To me it seemed like the main question to resolve with build status badges is more like when do we jam another edge case into `isBuildStatus` and when to just not use it, and I'm not sure I have more of an answer to that now.

**Reply by calebcartwright** [reply]
> Maybe we're just starting from different points and discovering that isBuildStatus is the most prevalent validator so we can say "default to isBuildStatus if you can" is enough of a clarification at this point.

Yeah good point. I think it's really more of an illustrative example. If you're using the `renderBuildStatus` helper, then you need some kind of schema strict enough to ensure that the helper can produce the right message value & color, and `isBuildStatus` is a good way to achieve that?

**Reply by PyvesB** [reply]
From a user perspective, I'm not convinced there's much value in making build status validation too strict. For example, if a CI returns "in progress", I would much rather have ![](https://img.shields.io/badge/build-in%20progress-informational) than ![](https://img.shields.io/badge/travis-invalid%20response%20data-critical). For us as maintainers, if we notice it we would immediately jump to the conclusion "we need to tweak the validator schema", but for some random person browsing someone else's repo, joining all the dots leading them to report a bug on the Shields.io project is a very long shot - most likely this transitional state will never be flagged up but will keep on popping up now and again and simply give the viewer the feeling that something has not been quite polished somewhere. That doesn't prevent us from having stricter validation on the test side though, so that we can notice more easily when a color or build message needs tweaking.

Generally speaking, I think that if we can render a badge that provides valuable information to the user, we should go ahead and render it. If the color is not perfect and defaults to the standard blue, it's not the end of the world in my opinion. Of course, I'm in agreement that for data that we can tell is blatantly wrong and doesn't convey useful information, e.g. ![](https://img.shields.io/badge/coverage---10%25-critical), there's no value in trying to render the badge.

**Reply by calebcartwright** [reply]
I guess I'm unclear on the degrees of "wrongness" for a badge. 

IIRC, an unknown status will result in a lightgrey/inactive message color and IMO a build badge like (I don't recall if `pass` is something already covered, just using it as an example of an unrecognized status with the helper)
![](https://img.shields.io/badge/build-pass-inactive) 

is just as blatantly wrong as the coverage badge
![](https://img.shields.io/badge/coverage---10%25-critical)

I also believe that one of the benefits of having some kind of minimal check within the schema beyond the presence of the key is that it can force the author/developer to do some up front thinking and review of the potential responses from the upstream service.

If we were to transition to just using `Joi.string()` for all build status badges and accepting that for all new badges I think we'd definitely have more incorrect badges.

**Reply by calebcartwright** [reply]
> but for some random person browsing someone else's repo, joining all the dots leading them to report a bug on the Shields.io project is a very long shot 

I do agree with you on this point though. I can never seem to find it when I need it, but there's been previous discussions about letting service classes override the value of this message and I suppose one option on the opposite end of the spectrum would be to change this to something very explicit like (but not literally :smile:) `Shields.io schema validation bug - please report at https://github.com/badges/shields/issues/new/choose`

**Comment by calebcartwright** [comment]
> I would say nexus is probably a good example of one where we are checking for something we don't really need to check at input time. It looks like nothing we are doing with latestRelease or latestSnapshot really requires them to be in a specific format.

Completely agreed, and that's why I picked it. We've had a few previous reports about `invalid response data` and had to adjust the regex accordingly. Huge benefit of relaxing this one is no longer having to battle regexes. 

> I think we do have consensus that not knowing what colour to render the badge is a condition we want to prevent and we should validate to guard against that

Yup, i think color would indeed fall under the "correct" umbrella we've been discussing here in addition to the message. So while yes, badges in the build status category would relatively have stricter schemas, the reason is because it's needed to ensure proper coloring of the badge.

That being said, there's obviously a coloring aspect. to version badges too. Many (most, all?) use the version helpers which also have some assumptions about the contents. I guess the difference here is that a version would have to be pretty wild (starting with something bonkers like `xa..`) to cause issues, and that doesn't seem like anywhere near as realistic a possibility as a build tool changing/adding a status variant 

https://github.com/badges/shields/blob/ff9273a958e8570dcb64f593f51b7743a1ca551f/services/color-formatters.js#L9-L23
